nav2_costmap_2d/include/nav2_costmap_2d/semantic_segmentation_layer.hpp

```diff
@@ -0,0 +1,161 @@
+  virtual ~SemanticSegmentationLayer() {}
+
+  /**
+   * @brief Initialization process of layer on startup
+   */
+  virtual void onInitialize();
+  /**
+   * @brief Update the bounds of the master costmap by this layer's update dimensions
+   * @param robot_x X pose of robot
+   * @param robot_y Y pose of robot
+   * @param robot_yaw Robot orientation
+   * @param min_x X min map coord of the window to update
+   * @param min_y Y min map coord of the window to update
+   * @param max_x X max map coord of the window to update
+   * @param max_y Y max map coord of the window to update
+   */
+  virtual void updateBounds(double robot_x, double robot_y, double robot_yaw, double* min_x,
+                            double* min_y, double* max_x, double* max_y);
+  /**
+   * @brief Update the costs in the master costmap in the window
+   * @param master_grid The master costmap grid to update
+   * @param min_x X min map coord of the window to update
+   * @param min_y Y min map coord of the window to update
+   * @param max_x X max map coord of the window to update
+   * @param max_y Y max map coord of the window to update
+   */
+  virtual void updateCosts(nav2_costmap_2d::Costmap2D& master_grid, int min_i, int min_j, int max_i,
+                           int max_j);
+
+  /**
+   * @brief Reset this costmap
+   */
+  virtual void reset();
+
+  virtual void onFootprintChanged();
+
+  /**
+   * @brief If clearing operations should be processed on this layer or not
+   */
+  virtual bool isClearable() { return true; }
+
+ private:
+ /**
+   * @brief Callback called when a poincloud and a segmentation are synced and a tf is available
+   * @param segmentation a pointer segmentation mask containing the class and confidence of each pixel
+   * @param pointcloud A pointer to the aligned pointcloud
+   */
+  void syncSegmPointcloudCb(
+    const std::shared_ptr<const vision_msgs::msg::SemanticSegmentation>& segmentation,
+    const std::shared_ptr<const sensor_msgs::msg::PointCloud2>& pointcloud);
+
+  ///< @brief subscriber to segmentation messages
+  std::shared_ptr<message_filters::Subscriber<vision_msgs::msg::SemanticSegmentation>>
+    semantic_segmentation_sub_;
+  ///< @brief subscriber to aligned pointclouds
+  std::shared_ptr<message_filters::Subscriber<sensor_msgs::msg::PointCloud2>> pointcloud_sub_;
+  ///< @brief subscriber to guarantee that transforms are available when pointclouds are received
+  std::shared_ptr<tf2_ros::MessageFilter<sensor_msgs::msg::PointCloud2>> pointcloud_tf_sub_;
+  ///< @brief subscriber to sync transforms, pointclouds and segmentations
+  std::shared_ptr<message_filters::TimeSynchronizer<vision_msgs::msg::SemanticSegmentation,
+                                                    sensor_msgs::msg::PointCloud2>>
+    segm_pc_sync_;
+
+  // debug publishers
+  ///< @brief publisher to publish segmentation messages when buffered
+  std::shared_ptr<rclcpp::Publisher<vision_msgs::msg::SemanticSegmentation>> sgm_debug_pub_;
```

Comment:
Remove `sgm_debug_pub_` `orig_pointcloud_pub_` `proc_pointcloud_pub_`, these are the same as the input topics, there's no reason to repeat them

Reply:
I think its worth leaving `proc_pointcloud_pub`. This publishes poinclouds with the segmentation `class` and `confidence` fields in each point. It may be useful for visualization. What do you think? This is subject to the `debug_topics` parameter anyway, which is False by default

Reply:
Messages are not published through this topic unless the parameter `debug_topics` is set to `true` in the plugin configuration. This parameter is `false` by default, so users should actually define it to get the messages. Do you still think it must be removed?

Reply:
OK, if you leave that last one that's OK